obs-ffmpeg: Several fixes allowing support of FFmpeg 5

pkviet opened the pull request:
### Description
The upgrade to FFmpeg master introduces
several issues:
- avcodec/version.h is no longer included in codec.h; avformat.h
doesn't carry the avcodec version any more.
- the new channel layout APIbreaks obs-ffmpeg-audio-encoders,
obs-ffmpeg-mux & obs-ffmpeg-output.

### Motivation and Context
Allow for a later upgrade to FFmpeg 5.

**Note:**
- I removed the convert_ff_channel_layout which is not compatible with
the new API; it is replaced by code which has compatibility with both
old and new channel layout API.
- I added an ffmpeg-mux fix for 4.0 channel layout.

### How Has This Been Tested?
obs-studio compiles (on win 10 x64, vs2019).
The FFmpeg audio encoders work (tested with streaming and recording).
The obs-ffmpeg-output works too.
FFmpeg version tested: 
- master branch on commit https://github.com/FFmpeg/FFmpeg/commit/944b8c96c90ae050606e638f77a15d6292209b91
(from April 25th 2022).
- This was also tested to work on ArchLinux and Manjaro (thx to @tytan652 ).

### Types of changes
- Bug fix (non-breaking change which fixes an issue) 

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

RytoEX reviewed (changes requested):
1. Do not use periods at the end of commit message subjects.
2. Use a verb as the first word of the commit message subject. "Fix for" uses "Fix" as a noun, not a verb.

Please clarify what version of FFmpeg you tested on: our current dependencies (4.4.1), 5.0.1, or something else. It does at least seem to build on CI, so hopefully it's fine with all current platforms. If the ifdefs are all checking against when the functions/features changed or were removed/added, then this should be fine.

pkviet commented:
> 1. Do not use periods at the end of commit message subjects.

fixed
> 2. Use a verb as the first word of the commit message subject. "Fix for" uses "Fix" as a noun, not a verb.

fixed
> 
> Please clarify what version of FFmpeg you tested on: our current dependencies (4.4.1), 5.0.1, or something else. 

- that was ffmpeg master at this commit : https://github.com/FFmpeg/FFmpeg/commit/944b8c96c90ae050606e638f77a15d6292209b91
(from April 25th 2022).
I didn't build FFmpeg on any branch.
> It does at least seem to build on CI, so hopefully it's fine with all current platforms. If the ifdefs are all checking against when the functions/features changed or were removed/added, then this should be fine.
- Indeed all the changes in this PR are ifdefd against ffmpeg libs versions where the API was broken, as documented in the commit messages (with links to the said commits breaking API).

pkviet commented:
**Update**

1. Commits text reduced to 72 chars (including links with hashes) & titles reduced to 50.
2. One new function removed per suggestion from Jim, in obs-ffmpeg-audio-encoders.c & replaced by lines 395-398.

RytoEX commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -495,14 +495,18 @@ static void create_audio_stream(struct ffmpeg_mux *ffm, int idx)
+#if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(59, 24, 100)
 	context->channel_layout =
 		av_get_default_channel_layout(context->channels);
-	//avutil default channel layout for 4 channels is 4.0 ; fix for quad
-	if (context->channels == 4)
-		context->channel_layout = av_get_channel_layout("quad");
```
Is this no longer an issue in older versions of libavcodec? How about newer versions (it's not included in the branch below)?

RytoEX reviewed (changes requested):
I've got a couple questions about these changes, but otherwise seems good.

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-audio-encoders.c:
```diff
@@ -263,12 +253,18 @@ static void *enc_create(obs_data_t *settings, obs_encoder_t *encoder,
 		av_opt_set(enc->context->priv_data, "aac_coder", "fast", 0);
 	}
-
+	char buf[256];
```
well spotted, i'll fix that.

pkviet replied:
resolved in new push

RytoEX reviewed (changes requested):
Don't line break URLs, even if they're longer than 72 characters.
![image](https://user-images.githubusercontent.com/624931/168376708-a999a420-ffca-447d-82da-15ddbf79beb8.png)

pkviet commented:
> Don't line break URLs, even if they're longer than 72 characters. ![image](https://user-images.githubusercontent.com/624931/168376708-a999a420-ffca-447d-82da-15ddbf79beb8.png)

fixed and pushed

RytoEX reviewed (approved):
I haven't built this myself, but CI built against the current deps (FFmpeg 4.4.1 with patches), and @pkviet says it builds against FFmpeg master, so it _should_ also build on any of the current FFmpeg 5 tags. If there's nothing functionally wrong here, this should be fine to merge.
